src/events.rs

```diff
@@ -372,29 +375,95 @@ impl<'js> EventEmitter<'js> {
     }
 }
 
-//TODO implement, abort(reason)
 #[rquickjs::class]
 #[derive(rquickjs::class::Trace)]
-struct AbortController {}
+pub struct AbortController<'js> {
+    signal: Class<'js, AbortSignal<'js>>,
+}
 
 #[rquickjs::methods]
-impl AbortController {
+impl<'js> AbortController<'js> {
     #[qjs(constructor)]
-    fn new() -> Self {
-        Self {}
+    pub fn new(ctx: Ctx<'js>) -> Result<Self> {
+        let abort_controller = Self {
+            signal: Class::instance(
+                ctx,
+                AbortSignal {
+                    aborted: false,
+                    reason: None,
+                },
+            )?,
+        };
+        Ok(abort_controller)
+    }
+
+    #[qjs(get)]
+    pub fn signal(&self) -> Class<'js, AbortSignal<'js>> {
+        self.signal.clone()
+    }
+
+    pub fn abort(
+        this: This<Class<'js, Self>>,
+        reason: Opt<Value<'js>>,
+    ) -> Result<Class<'js, Self>> {
+        if reason.0.is_some() {
+            this.0.borrow_mut().signal.borrow_mut().set_reason(reason)?;
+        } else {
+            // TODO store DOMException as reason
+            this.0.borrow_mut().signal.borrow_mut().set_reason(reason)?;
+        }
+
+        Ok(this.0)
     }
 }
 
 //TODO implement static methods abort() and timeout(miliseconds)
 #[rquickjs::class]
-#[derive(rquickjs::class::Trace)]
-struct AbortSignal {}
+pub struct AbortSignal<'js> {
+    aborted: bool,
+    reason: Option<Value<'js>>,
+}
+
+impl<'js> Trace<'js> for AbortSignal<'js> {
+    fn trace<'a>(&self, tracer: Tracer<'a, 'js>) {
+        if let Some(reason) = &self.reason {
+            tracer.mark(reason);
+        }
+    }
+}
 
 #[rquickjs::methods]
-impl AbortSignal {
+impl<'js> AbortSignal<'js> {
     #[qjs(constructor)]
-    fn new() -> Self {
-        Self {}
+    pub fn new() -> Self {
+        Self {
+            aborted: false,
+            reason: None,
+        }
+    }
+
+    #[qjs(get)]
+    pub fn aborted(&self) -> bool {
+        self.aborted
+    }
+
+    pub fn set_reason(&mut self, reason: Opt<Value<'js>>) -> Result<Option<Value<'js>>> {
```

Comment:
Use `#[qjs(skip)]` or move to another `impl`

Reply:
Should this also set aborted to true?

Reply:
> Should this also set aborted to true?

I wasn't sure as the spec didn't mention the abort controller setting `aborted` to true. I assume maybe whatever observes that the `reason` has changed would then set `aborted` to true. I could set it to true for now till that is implemented

Reply:
Also what does `qjs(skip)` do? I couldn't find any documentation for it 👀

Reply:
`qjs(skip)` will not expose the function to JS user land.